db/repair_bucket.go

```diff
@@ -0,0 +1,227 @@
+type DocTransformer func(docId string, originalCBDoc []byte) (transformedCBDoc []byte, transformed bool, err error)
+
+// A RepairBucket struct is the main API entrypoint to call for repairing documents in buckets
+type RepairBucket struct {
+	DryRun     bool // If true, will only output what changes it *would* have made, but not make any changes
+	Bucket     base.Bucket
+	RepairJobs []DocTransformer
+}
+
+func NewRepairBucket(bucket base.Bucket) *RepairBucket {
+	return &RepairBucket{
+		Bucket: bucket,
+	}
+}
+
+func (r *RepairBucket) SetDryRun(dryRun bool) *RepairBucket {
+	r.DryRun = dryRun
+	return r
+}
+
+func (r *RepairBucket) AddRepairJob(repairJob DocTransformer) *RepairBucket {
+	r.RepairJobs = append(r.RepairJobs, repairJob)
+	return r
+}
+
+func (r *RepairBucket) InitFrom(params RepairBucketParams) *RepairBucket {
+
+	r.SetDryRun(params.DryRun)
+	for _, repairJobParams := range params.RepairJobs {
+		switch repairJobParams.RepairJobType {
+		case RepairRevTreeCycles:
+			r.AddRepairJob(RepairJobRevTreeCycles)
+		}
+	}
+
+	return r
+}
+
+func (r RepairBucket) RepairBucket() (results []RepairBucketResult, err error) {
+
+	options := Body{"stale": false, "reduce": false}
+	options["startkey"] = []interface{}{true}
+
+	vres, err := r.Bucket.View(DesignDocSyncHousekeeping, ViewImport, options)
+	if err != nil {
+		return results, err
+	}
+
+	for _, row := range vres.Rows {
+		rowKey := row.Key.([]interface{})
+		docid := rowKey[1].(string)
+		key := realDocID(docid)
+
+		err = r.Bucket.Update(key, 0, func(currentValue []byte) ([]byte, error) {
+			// Be careful: this block can be invoked multiple times if there are races!
+			if currentValue == nil {
+				return nil, couchbase.UpdateCancel // someone deleted it?!
+			}
+
+			updatedDoc, shouldUpdate, repairJobs, err := r.TransformBucketDoc(key, currentValue)
+			if err != nil {
+				return nil, err
+			}
+
+			switch shouldUpdate {
+			case true:
+
+				backupOrDryRunDocId, err := r.WriteRepairedDocsToBucket(key, currentValue, updatedDoc)
+				if err != nil {
+					base.LogTo("CRUD", "Repair Doc (dry_run=%v) Writing docs to bucket failed with error: %v.  Dumping raw contents.", r.DryRun, err)
+					base.LogTo("CRUD", "Original Doc before repair: %s", currentValue)
+					base.LogTo("CRUD", "Updated doc after repair: %s", updatedDoc)
+				}
+				if r.DryRun {
+					base.LogTo("CRUD", "Repair Doc: dry run result available in Bucket Doc: %v (auto-deletes in 1 hour)", backupOrDryRunDocId)
+				} else {
+					base.LogTo("CRUD", "Repair Doc: Doc repaired, original doc backed up in Bucket Doc: %v (auto-deletes in 1 hour)", backupOrDryRunDocId)
+				}
+
+				results = append(results, RepairBucketResult{DocId: key, RepairJobTypes: repairJobs})
+
+				if r.DryRun {
+					return nil, couchbase.UpdateCancel
+				} else {
+					return updatedDoc, nil
+				}
+			default:
+				return nil, couchbase.UpdateCancel
+			}
+
+		})
+
```

Comment:
Do error checking for the error returned by Update.

If that's returning UpdateCancel for dry run, can move the 'update complete' logging after the error is checked.

Reply:
Done